bin/cml-tensorboard-dev.js

```diff
@@ -32,51 +33,52 @@ const run = async (opts) => {
   await fs.writeFile(`${path}/uploader-creds.json`, credentials);
 
   // launch  tensorboard on background
-  const tb_path = await exec('which tensorboard');
   const help = await exec('tensorboard dev upload -h');
   const extra_params_found =
     (name || description) && help.indexOf('--description') >= 0;
   const extra_params = extra_params_found
     ? `--name "${name}" --description "${description}"`
     : '';
-  const command = `python -u ${tb_path} dev upload --logdir ${logdir} ${extra_params}`;
 
-  const proc = spawn(command, { detached: true, shell: true });
-
-  proc.stderr.on('data', (data) => {
-    data && console.error(data.toString('utf8'));
+  const command = `tensorboard dev upload --logdir ${logdir} ${extra_params}`;
+  const stdout_path = tempy.file({ extension: 'log' });
+  const stdout_fh = await fs.open(stdout_path, 'a');
+  const stderr_path = tempy.file({ extension: 'log' });
+  const stderr_fh = await fs.open(stderr_path, 'a');
+  const proc = spawn(command, [], {
+    detached: true,
+    shell: true,
+    stdio: ['ignore', stdout_fh, stderr_fh]
   });
 
-  let output = '';
-  proc.stdout.on('data', async (data) => {
-    if (data) {
-      console.error(data.toString('utf8'));
+  proc.unref();
+  proc.on('exit', (code) => {
+    throw new Error(`Tensorboard process exited with code ${code}`);
+  });
 
-      output += data.toString('utf8');
+  // reads stdout every 5 secs to find the tb uri
+  setInterval(async () => {
+    const stdout_data = await fs.readFile(stdout_path, 'utf8');
+    const regex = /(https?:\/\/[^\s]+)/;
+    const matches = stdout_data.match(regex);
 
-      const uri_index = output.indexOf('https://', 0);
-      if (uri_index > -1) {
-        output = output.substring(uri_index, output.length - 1);
+    if (matches.length) {
+      let output = matches[0];
 
-        if (md) output = `[${title || name}](${output})`;
+      if (md) output = `[${title || name}](${output})`;
 
-        if (!file) print(output);
-        else await fs.writeFile(file, output);
+      if (!file) print(output);
+      else await fs.appendFile(file, output);
 
-        process.exit(0);
-      }
+      stdout_fh.close() && stderr_fh.close();
```

Comment:
it looks like you can end up in a situation when these two won't be closed.
what if `if (matches.length)` is false?

Reply:
It will be handled and closed by the 1min execution timeout

Reply:
it is still better to close all the files properly.

Reply:
they are closed properly. If matches.length does not happen is because there is an error in tb that makes impossible to extract the url. Our 1min exception timeout will close the descriptors.